File: apps/web/src/realtime/events.ts
export type Speaker = "Ethan" | "Chelsie" | "Aiden";

export type GatewayMetrics = {
  vad_end_to_commit_ms?: number | null;
  commit_to_qwen_first_audio_ms?: number | null;
  qwen_first_audio_to_livekit_first_frame_ms?: number | null;
  speech_end_to_first_assistant_egress_ms?: number | null;
  speech_start_to_commit_ms?: number | null;
  turn_total_ms?: number | null;
  queue_depth_at_first_frame_ms?: number | null;
  assistant_queue_depth_ms?: number | null;
  interrupt_clear_ms?: number | null;
  commit_to_first_audio_played_ms?: number | null;
  mic_to_first_transcript_ms?: number | null;
  commit_to_first_transcript_ms?: number | null;
  commit_to_first_text_ms?: number | null;
  commit_to_first_audio_delta_ms?: number | null;
  commit_to_first_livekit_egress_ms?: number | null;
  full_response_ms?: number | null;
};

export type GatewayRollups = Record<string, number | null>;

export type GatewayCounters = {
  error_count?: number | null;
  interruption_count?: number | null;
  duplicate_commit_count?: number | null;
  stale_output_drop_count?: number | null;
};

type GatewayEventBase = {
  session_id?: string | null;
  participant_id?: string | null;
  participant_identity?: string | null;
  room?: string | null;
  turn_id?: string | null;
  input_epoch?: number | null;
  interrupt_epoch?: number | null;
  output_version?: number | null;
};

export type GatewayInboundEvent =
  | (GatewayEventBase & {
      type: "session.ready";
      session_id: string;
      backend?: "qwen_realtime_audio" | "qwen_chat_stream_audio" | "text_only";
      degraded_reason?: string | null;
    })
  | (GatewayEventBase & { type: "turn.started"; speech_duration_ms?: number | null })
  | (GatewayEventBase & {
      type: "turn.committed";
      speech_duration_ms?: number | null;
      silence_duration_ms?: number | null;
      input_audio_ms?: number | null;
    })
  | (GatewayEventBase & { type: "transcript.delta"; text: string })
  | (GatewayEventBase & { type: "assistant.text.delta"; text: string })
  | (GatewayEventBase & {
      type: "assistant.audio_started";
      sample_rate?: number | null;
      channels?: number | null;
      format?: string | null;
    })
  | (GatewayEventBase & {
      type: "assistant.audio.metadata";
      sample_rate?: number | null;
      channels?: number | null;
      format?: string | null;
      audio_base64_length?: number | null;
    })
  | (GatewayEventBase & { type: "assistant.done" })
  | (GatewayEventBase & { type: "assistant.interrupted" })
  | (GatewayEventBase & {
      type: "metrics.update";
      metrics: GatewayMetrics;
      timestamps?: Record<string, number | null>;
      rollups?: GatewayRollups;
      counters?: GatewayCounters;
      reconnects?: Record<string, number | null>;
      last_reconnect_reason?: string | null;
    })
  | (GatewayEventBase & { type: "room.busy"; code: "ROOM_BUSY"; message: string })
  | (GatewayEventBase & { type: "error"; code: string; message: string });

File: apps/gateway/src/realtime/events.py
from __future__ import annotations

from typing import Literal, TypeAlias, TypedDict


PublicBackend = Literal[
    "qwen_realtime_audio",
    "qwen_chat_stream_audio",
    "text_only",
]


class GatewayEventBase(TypedDict, total=False):
    session_id: str
    participant_id: str
    participant_identity: str
    room: str
    turn_id: str
    input_epoch: int
    interrupt_epoch: int
    output_version: int
    backend: PublicBackend
    reason: str
    rollups: dict[str, float | int | None]


class SessionReadyEvent(GatewayEventBase, total=False):
    type: Literal["session.ready"]
    degraded_reason: Literal["non_realtime_audio_backend"] | None


class TurnStartedEvent(GatewayEventBase, total=False):
    type: Literal["turn.started"]
    speech_duration_ms: float | None


class TurnCommittedEvent(GatewayEventBase, total=False):
    type: Literal["turn.committed"]
    speech_duration_ms: float | None
    silence_duration_ms: float | None
    input_audio_ms: float | None


class TranscriptDeltaEvent(GatewayEventBase):
    type: Literal["transcript.delta"]
    text: str


class AssistantTextDeltaEvent(GatewayEventBase):
    type: Literal["assistant.text.delta"]
    text: str


class AssistantAudioStartedEvent(GatewayEventBase, total=False):
    type: Literal["assistant.audio_started"]
    sample_rate: int | None
    channels: int | None
    format: str | None


class AssistantAudioMetadataEvent(GatewayEventBase, total=False):
    type: Literal["assistant.audio.metadata"]
    sample_rate: int | None
    channels: int | None
    format: str | None
    audio_base64_length: int | None


class AssistantDoneEvent(GatewayEventBase):
    type: Literal["assistant.done"]


class AssistantInterruptedEvent(GatewayEventBase):
    type: Literal["assistant.interrupted"]


class MetricsUpdateEvent(GatewayEventBase, total=False):
    type: Literal["metrics.update"]
    metrics: dict[str, float | int | None]
    timestamps: dict[str, float | int | None]
    reconnects: dict[str, int]
    last_reconnect_reason: str | None


class ErrorEvent(GatewayEventBase):
    type: Literal["error"]
    code: str
    message: str


class RoomBusyEvent(GatewayEventBase):
    type: Literal["room.busy"]
    code: Literal["ROOM_BUSY"]
    message: str


class AssistantAudioDeltaEvent(GatewayEventBase, total=False):
    type: Literal["assistant.audio.delta"]
    audio_base64: str
    sample_rate: int | None
    channels: int | None
    format: str | None


class InputSpeechTruncatedEvent(GatewayEventBase, total=False):
    type: Literal["input.speech.truncated"]
    max_turn_ms: int
    input_audio_ms: float | None
    reason: str


PublicGatewayEvent: TypeAlias = (
    SessionReadyEvent
    | TurnStartedEvent
    | TurnCommittedEvent
    | TranscriptDeltaEvent
    | AssistantTextDeltaEvent
    | AssistantAudioStartedEvent
    | AssistantAudioMetadataEvent
    | AssistantDoneEvent
    | AssistantInterruptedEvent
    | MetricsUpdateEvent
    | ErrorEvent
    | RoomBusyEvent
    | InputSpeechTruncatedEvent
)

RuntimeGatewayEvent: TypeAlias = PublicGatewayEvent | AssistantAudioDeltaEvent
